Prefetching (WIP)

dmitrym0 opened the pull request:
# ~~**_This is a work in progress. Please don't merge.**_~~ please see updates further down.

I'm going to be away on leave until May, so this is a placeholder for tracking this feature.
# Summary

The intent of the prefetching feature is to allow seamless inter-spine transitions. In cloud reader configuration, spines are fetched "just in time", right when the reader is about to display them. Depending on content size and network throughput this introduces a delay that is interrupts the reading flow. This feature aims to solve this problem by intelligently prefetching necessary spine items and rendering them when requested by the reader.
# Current status

As of Friday, February 27, 2015, this pull request contains a proof of contept implementation for reflowable and fixed layout views. Reflowable views are prefetched and rendered correctly in the background. Fixed layout views are working, but not as well as reflowable views. There's a noticeable delay when fixed layout views are displayed, it probably means that the fixed layouts views haven't completed full rendering.
# Next steps
1. Refactor code. _reader_view_ doesnt necessarily need to know about anything caching related.
2. Refactor _one_page_view_ to support correct rendering order so that when the iframe is shown everything is rendered.
3. Come up with a caching strategy. Currently spines persist forever.
4. Determine how prefetching interacts with other Readium features. Since the spine is essentially loaded but not displayed does that mean that Media overlays will start playing automatically?
5. Interaction with other types of views (continuous)?

Commit:
Prefetching.

Resolving merge conflicts.

more merge fixes

Commit:
Single page fixed layout views work..ish

Commit:
Refactored triggering of events for cached items

Commit:
Cleanly determine which spines are currently cached

Commit:
Prefetch previous and next spines based on the displaying view type

Commit:
Only store cached views once they are fully loaded. Also, store current view when its fully loaded

Commit:
Reformat code slightly.

Commit:
Remove event handlers when the view comes off the screen

Commit:
Check to see that we're not creating duplicates when adding cached views.

Commit:
Attempt to hide the inactive views in order to solve the duplicate rendering problem

Commit:
Make sure that previous spine item is valid (eg: at the beginning of the book) before loading it

Commit:
Add simple expiration strategy for cached views

dmitrym0 commented:
# Current state

- fetches "previous" and "next" pages while the "current" page is being perused. In this context a "page" may mean two spine items (fixed layout, synthetic spread) or a single reflowable spine item.
- evicts stale "pages" via a simple strategy (the distance between current spine item and the cached view.

# To Do

- there's a fairly significant bug where if the user advances through the pages rapidly the reader view loses track of visible spine items and two or more spines may overlap on screen.
- possibly extract some of the added functionality into a utility class, not everything needs to be in reader_view.
- **Testing**:
  - interaction with media overlays
  - scroll views?

Commit:
Initial refactoring, create cache_manager class and move some of the obvious canditates in there.

Commit:
Basic functionality with view management residing outside of reader_view

Commit:
Proxy the document load events to the reader. Everything seems to work, prefetching currently disabled.

Commit:
Fixes. Both reflowable and fixed layout views work

Commit:
Rename CacheManager to ViewManager

Commit:
Temporarily ignore font sizes in reflowable view.

Commit:
Persist the iframe references inside the view manager (this is so we can remove public access to iframes from the views)

Commit:
Remove reader dependency inside the view classes; instead pass a settings instance

Commit:
Refactor out references to the reader inside the views, pass in options inside

dmitrym0 commented:
## Current status: ready to be reviewed & tested.

# Summary of changes:

1. View management (creation, caching, expiration) is delegated to the `view_manager.js` class. This allowed to me localize changes related to multiple views to this class only and leave `reader_view.js` largely unchanged.
2. Add functionality to the view classes (`one_page_view` and `reflowable_view`) that allow these to be instantiated without being visible.
3. Tiny refactoring to remove circular dependency on `reader_view` from within the view classes.

# Next steps:

1. Code review & feedback
2. Testing

Commit:
Clean up a typo

Commit:
Propagate cached view information to the subviews of the fixed view. This fixes a bug where multiple FXL views are visible.

Commit:
Reenable cache expiration

Commit:
Propagate viewer settings to the invidividual views

Commit:
Properly set view settingsi n reflowable view

Commit:
Add a few comments to the view manager

Commit:
Moved viewTypeForView() function into the section for public functions.

Commit:
Fix a typo, passing view settings into the child view.

Commit:
Remove dead code, exposing internal implementation of the views is no longer necessary

Commit:
Remove more dead code.

dmitrym0 commented:
# Current status: Dev complete

As of the last commit, I'd like to declare this feature dev complete. About time :smile:.

# Next steps:

- I would still like a code review. Maybe @danielweck or @JCCR?
- **Testing**. *"It works on the developer machine"* is not good enough, but I haven't observed any major issues. However I haven't tested the following:
  - media overlays
  - scroll views.
- After @rkwright's go ahead it can be merged into develop.

Commit:
Fix iframe names based on Jonahs feedback. iFrame names are now unique

Commit:
Minor changes to the code suggested by the linter.

danielweck commented:
Looking at the code diff, `git merge develop` in the `feature/prefetch` branch should resolve pretty straight-forwardly. Unfortunately personally I will not be able to test in the coming days.

ViewManager can easily be wrapped with AMD `define([], function() { };`:
https://github.com/readium/readium-shared-js/pull/160/files#diff-ba772a19972832ecbcfee79bf809d85b

Commit:
AMD'ify view_manager. Require it in reader_view.

Commit:
Move view type constants to view manager.

Commit:
Whoops, this should not be here...

JCCR commented:
@dmitrym0 @danielweck 
Updated to match `develop`.

Commit:
Make the viewManager not need to hold a reference of a view's $iframe in order to replay view lifecycle events.

Commit:
Fix issues introduced with the view manager concerning hiding/showing iframe and epub content in reflowable views.

Commit:
Correct the usage of viewerSettings, and fix the creation of scroll views.
(I didn't want to introduce a circular dependancy on ViewManager and ReaderView but I have no choice until the ScrollView's circular dependency is also removed)

Commit:
Looks like actually using a ScrollView with the ViewManager refactoring makes them broken...

(Also this commit "cleans up" whitespace, sorry about this, but we have to normalize...)

Commit:
Introduce more "aggressive" view cleanup. This is an attempt to isolate a memory leak.

danielweck commented:
@JCCR Agree about disabling prefetch in the `continuous` scroll view, as there is already a "stack" of iframes / loaded documents (okay to keep in the single `document` scroll).

Also, here's a new issue about the annotations plugin potential memory leaks:
https://github.com/readium/readium-js-viewer/issues/386

JCCR commented:
During my testing I found out that some of the cached views don't "let go" of their iframe references. This issue can be clearly spotted by looking at the DOM tree while hitting the next page button in a quick succession. After doing this there will be many iframes in the DOM, more than there should be. This looks like a race condition with the new view manager refactoring.
I'm using "accessible epub3" as the test book.

While looking at this I also spotted a large amount of growth with the JS + DOM nodes heap snapshots when switching back and forth between two of the same spine items. When doing a comparison of before and after I noticed the biggest items on the list were objects on the EpubAnnotations namespace, and some of Backbone's internals. After disabling the annotations plugin and repeating the same actions the memory growth lowered substantially, going from 5-10mb per spine item switch, to just 0.1-1mb

JCCR commented:
## Results from testing
#### Issues with scroll view (document mode):

Right now the containers for the cached views are being stacked vertically and on top of each other, causing the one that should be front and visible to not show up regular conditions. The idea for laying out the containers for cached views is to have a horizontal stack or strip of the views. This needs to be fixed or implemented in the CSS styling for the scrollview like it was for reflowable view.

#### Issues with continuous scroll view:
It's broken, but this is because they try to do the same thing. The solution seems to be to disabling this prefetching work when dealing with this type of scroll view.

#### Memory leaks?
Besides the clear and concise bug with some iframes being left behind in the DOM due to a race condition in view manager, there might be further issues with not properly cleaning up references to iframes, other DOM nodes, event bindings, etc. These issues are not introduced with the prefetching itself, as they seem to be a more general problem with the codebase, but they are easier to spot with prefetching.

#### Non linear spine items support
Cached views are being created by making calculations with spine item indices, as shown here: https://github.com/readium/readium-shared-js/pull/160/files?diff=unified#diff-ba772a19972832ecbcfee79bf809d85bR119
This needs to instead take into account of non linear spine item with calculations based on the readium-js `spine` abstraction. `reader.spine().nextItem()` and so on..

jdempcy commented:
I ran into an issue where autoplay audio/video from spine items which are prefetched starts playing before going to that page.

For instance, say there is autoplay video in spine item 3. When navigating from spine item 1 to 2, spine item 3 will be prefetched and the video will automatically start playing, even though I'm still viewing spine item 2.

danielweck commented:
@jdempcy yep, and this is also a problem in the general case (no prefetch):
https://github.com/readium/readium-shared-js/issues/78

dmitrym0 commented:
Closing, stale.
